File: src/main/python/CheckPermutation.py
'''
Check Permutation:
Given two strings, write a method to decide if one is a permutation of the
other.
'''

from typing import DefaultDict


def checkPermutation(s1:str,s2:str)->bool:
    if len(s1)!=len(s2):return False

    mem = DefaultDict(int) 
    for c in s1:
        mem[c] += 1
    
    for c in s2:
        if mem[c] == 0: return False
        else:mem[c] -= 1
    
    return True

def checkPermutation2(s1:str,s2:str)->bool:
    return sorted(s1) == sorted(s2)

if __name__ == '__main__':
    assert checkPermutation('abcd','dacb') == True
    assert checkPermutation('abcd','dacw') == False
    assert checkPermutation('','dacw') == False

    assert checkPermutation2('abcd','dacb') == True
    assert checkPermutation2('abcd','dacw') == False
    assert checkPermutation2('','dacw') == False

File: src/main/C++/checkPermutation.cpp
#include<bits/stdc++.h>

using namespace std;

bool checkPermutation(string s1, string s2)
{
    sort(s1.begin(),s1.end());
    sort(s2.begin(),s2.end());

    return s1 == s2;
}

bool checkPermutation2(string s1,string s2)
{
    if (s1.length()!=s2.length())
        return false;

    map<char,int> mem;
    for (auto& c:s1)
        mem[c] +=1;

    for (auto& c:s2){
        if(mem[c]==0)
            return false;
        else
            mem[c] -= 1;
    }

    return true;

}

int main()
{
    assert ( checkPermutation("abcd","dacb") == true );
    assert ( checkPermutation("abcd","dacbe") == false );
    assert ( checkPermutation("abcd","da") == false );

    assert ( checkPermutation2("abcd","dacb") == true );
    assert ( checkPermutation2("abcd","dacbe") == false );
    assert ( checkPermutation2("abcd","da") == false );
    return 0;
}